File: Leetcode/0083-remove-duplicates-from-sorted-list/solution.go
package main

import "fmt"

// 0083. Remove Duplicates from Sorted List
// Time: O(n), Space: O(1)

type ListNode struct {
	Val  int
	Next *ListNode
}

func deleteDuplicates(head *ListNode) *ListNode {
	cur := head
	for cur != nil && cur.Next != nil {
		if cur.Next.Val == cur.Val {
			cur.Next = cur.Next.Next
		} else {
			cur = cur.Next
		}
	}
	return head
}

func toList(arr []int) *ListNode {
	if len(arr) == 0 {
		return nil
	}
	h := &ListNode{Val: arr[0]}
	c := h
	for _, v := range arr[1:] {
		c.Next = &ListNode{Val: v}
		c = c.Next
	}
	return h
}
func toSlice(h *ListNode) []int {
	out := []int{}
	for h != nil {
		out = append(out, h.Val)
		h = h.Next
	}
	return out
}
func eq(a, b []int) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

func main() {
	passed, failed := 0, 0
	test := func(name string, got, exp []int) {
		if eq(got, exp) {
			fmt.Printf("PASS: %s\n", name)
			passed++
		} else {
			fmt.Printf("FAIL: %s got=%v\n", name, got)
			failed++
		}
	}
	cases := []struct {
		name string
		in   []int
		want []int
	}{
		{"Example 1", []int{1, 1, 2}, []int{1, 2}},
		{"Example 2", []int{1, 1, 2, 3, 3}, []int{1, 2, 3}},
		{"Empty", []int{}, []int{}},
		{"Single", []int{5}, []int{5}},
		{"All same", []int{1, 1, 1}, []int{1}},
		{"No dup", []int{1, 2, 3}, []int{1, 2, 3}},
		{"Long", []int{0, 0, 0, 1, 2, 2, 3, 3, 3, 4}, []int{0, 1, 2, 3, 4}},
	}
	for _, c := range cases {
		test(c.name, toSlice(deleteDuplicates(toList(c.in))), c.want)
	}
	fmt.Printf("\nResults: %d passed, %d failed, %d total\n", passed, failed, passed+failed)
}

File: Leetcode/0083-remove-duplicates-from-sorted-list/solution.py
from typing import Optional, List


class ListNode:
    def __init__(self, val: int = 0, next: Optional['ListNode'] = None):
        self.val = val
        self.next = next


class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        """Time O(n), Space O(1)."""
        cur = head
        while cur and cur.next:
            if cur.next.val == cur.val:
                cur.next = cur.next.next
            else:
                cur = cur.next
        return head


def to_list(arr):
    if not arr: return None
    h = ListNode(arr[0]); c = h
    for v in arr[1:]: c.next = ListNode(v); c = c.next
    return h


def to_arr(h):
    out = []
    while h: out.append(h.val); h = h.next
    return out


if __name__ == "__main__":
    sol = Solution()
    passed = failed = 0
    def test(name, got, expected):
        global passed, failed
        if got == expected: print(f"PASS: {name}"); passed += 1
        else: print(f"FAIL: {name}"); failed += 1
    cases = [
        ("Example 1", [1, 1, 2], [1, 2]),
        ("Example 2", [1, 1, 2, 3, 3], [1, 2, 3]),
        ("Empty", [], []),
        ("Single", [5], [5]),
        ("All same", [1, 1, 1], [1]),
        ("No dup", [1, 2, 3], [1, 2, 3]),
        ("Long", [0, 0, 0, 1, 2, 2, 3, 3, 3, 4], [0, 1, 2, 3, 4]),
    ]
    for n, a, exp in cases: test(n, to_arr(sol.deleteDuplicates(to_list(a))), exp)
    print(f"\nResults: {passed} passed, {failed} failed, {passed + failed} total")
